File: PTASCustomSearches/Source/CustomSearches/CustomSearchesBE/CustomSearchesEFLibrary/Gis/Model/LayerSource.cs
using System;
using System.Collections.Generic;

namespace CustomSearchesEFLibrary.Gis.Model
{
    public partial class LayerSource
    {
        public LayerSource()
        {
            MapRenderer = new HashSet<MapRenderer>();
            MobileLayerRenderers = new HashSet<MobileLayerRenderer>();
        }

        public int LayerSourceId { get; set; }
        public string LayerSourceName { get; set; }
        public string GisLayerName { get; set; }
        public string Metadata { get; set; }
        public string LayerSourceAlias { get; set; }
        public string Jurisdiction { get; set; }
        public string Organization { get; set; }
        public string Description { get; set; }
        public bool IsParcelSource { get; set; }
        public int DefaultMinZoom { get; set; }
        public int DefaultMaxZoom { get; set; }
        public int DefaultLabelMinZoom { get; set; }
        public int DefaultLabelMaxZoom { get; set; }
        public bool IsVectorLayer { get; set; }
        public string DataSourceUrl { get; set; }
        public int TileSize { get; set; }
        public string DefaultMapboxLayer { get; set; }
        public string NativeMapboxLayers { get; set; }
        public string DefaultLabelMapboxLayer { get; set; }
        public bool HasOfflineSupport { get; set; }
        public bool HasMobileSupport { get; set; }
        public bool HasOverlapSupport { get; set; }
        public bool IsBlobPassThrough { get; set; }
        public string DbTableName { get; set; }
        public string EmbeddedDataFields { get; set; }
        public string OgrLayerData { get; set; }
        public bool ServeFromFileShare { get; set; }

        public virtual ICollection<MapRenderer> MapRenderer { get; set; }
        public virtual ICollection<MobileLayerRenderer> MobileLayerRenderers { get; set; }
    }
}

File: PTASPortals/Source/mapboxcontrol/src/services/map/model/LayerSource.ts
// LayerSource.ts
/**
 * Copyright (c) King County. All rights reserved.
 * @packageDocumentation
 */

import { camelCase } from 'lodash';
import {
  VectorSource,
  RasterSource,
  RasterLayer,
  FillLayer,
  LineLayer,
  SymbolLayer,
  CircleLayer,
} from 'mapbox-gl';
import { mapColumnType } from 'utils';
import { ApiColumnType, DataSetColumnType } from './renderer/rendererDataset';

export interface EmbeddedDataFields {
  FieldName: string;
  FieldType: DataSetColumnType;
}

export interface EmbeddedMapped {
  FieldName: string;
  FieldType: ApiColumnType;
}

/**
 * Map Sources vector and raster
 */
export class MapSource
  implements
    Omit<mapboxgl.VectorSource, 'type'>,
    Omit<mapboxgl.RasterSource, 'type'> {
  type: 'vector' | 'raster';
  tiles: string[];
  maxzoom: number | undefined;

  constructor(
    fields: Pick<
      LayerSource,
      'dataSourceUrl' | 'layerSourceType' | 'defaultMaxZoom'
    >
  ) {
    this.tiles = [fields.dataSourceUrl];
    this.type = fields.layerSourceType;
    this.maxzoom = fields.defaultMaxZoom;
  }
}

type LayerMetadata = {
  metadata: {
    scaleMinZoom?: number;
    scaleMaxZoom?: number;
    isSystemRenderer: boolean;
  };
};

export type SymbolPtasLayer = Omit<SymbolLayer, 'metadata'> & LayerMetadata;

export type PtasLayer =
  | SymbolPtasLayer
  | (Omit<FillLayer, 'metadata'> & LayerMetadata)
  | (Omit<LineLayer, 'metadata'> & LayerMetadata)
  | (Omit<RasterLayer, 'metadata'> & LayerMetadata)
  | (Omit<CircleLayer, 'metadata'> & LayerMetadata);

export type LayerSourceRes = Omit<LayerSource, 'embeddedDataFields'> & {
  embeddedDataFields: EmbeddedMapped[] | null;
};

/**
 * Layer configuration model, with map source
 */
export class LayerSource {
  hasOverlapSupport: boolean;
  nativeMapboxLayers: PtasLayer[];
  layerSourceId: number;
  layerSourceName: string;
  layerSourceAlias: string;
  layerSourceType: 'vector' | 'raster';
  jurisdiction: string;
  organization: string;
  description: string;
  isParcelSource: boolean;
  isVectorLayer: number;
  dataSourceUrl: string;
  defaultMapboxLayer: PtasLayer;
  mapSource: VectorSource | RasterSource;
  embeddedDataFields: EmbeddedDataFields[];
  defaultMaxZoom: number | undefined;
  metadata: {
    id: string;
    name: string;
    org: string;
    subject: string;
    dataType: string;
    abstract: string;
    kw: string;
  } | null = null;

  constructor(fields: LayerSourceRes) {
    this.hasOverlapSupport = fields.hasOverlapSupport;
    this.layerSourceId = fields.layerSourceId;
    this.layerSourceName = fields.layerSourceName;
    this.layerSourceAlias = fields.layerSourceAlias;
    this.jurisdiction = fields.jurisdiction;
    this.organization = fields.organization ?? '';
    this.description = fields.description;
    this.isParcelSource = fields.isParcelSource;
    this.isVectorLayer = fields.isVectorLayer;
    this.dataSourceUrl = `${process.env.REACT_APP_MAP_TILE_SERVICE_HOST}${fields.dataSourceUrl}`;
    this.defaultMaxZoom = fields.defaultMaxZoom;

    this.defaultMapboxLayer = {
      ...fields.defaultMapboxLayer,
      layout: {
        ...(fields.defaultMapboxLayer?.layout ?? {}),
        visibility: 'visible',
      },
    };
    this.embeddedDataFields =
      fields.embeddedDataFields?.map((val) => ({
        FieldName: val.FieldName,
        FieldType: mapColumnType(val.FieldType),
      })) ?? [];

    this.layerSourceType = fields.isVectorLayer ? 'vector' : 'raster';
    this.mapSource = new MapSource({
      dataSourceUrl: this.dataSourceUrl,
      layerSourceType: this.layerSourceType,
      defaultMaxZoom: this.defaultMaxZoom,
    });

    this.nativeMapboxLayers = fields.nativeMapboxLayers ?? [];

    if (fields.metadata) {
      this.metadata = (Object.entries(fields.metadata).reduce(
        (prev, [k, v]) => {
          return {
            ...prev,
            [camelCase(k)]: v,
          };
        },
        {}
      ) as unknown) as LayerSource['metadata'];
    }
  }
}
